In [1]:
import sys, os
sys.path.append(os.path.abspath("/Users/DavidHolzwarth/Uni/EPFL/bachelors-thesis"))

from graph_viz import visualize_delegation_graph
from LP import invert_graph, resolve_delegations_LP, set_up_LP, solve_LP
from iterative import iterate_delegations


delegations = {
    "A": {"B": 0.9999,
          "C": 0.0001},
    "B": {"A": 1}}

nodes = ["A", "B", "C"]

delegations = invert_graph(delegations)

model, _, _ = set_up_LP(delegations, nodes)

%timeit solve_LP(model)
print(resolve_delegations_LP(delegations, nodes))
%timeit iterate_delegations(delegations, nodes)

print(iterate_delegations(delegations, nodes))

/opt/anaconda3/envs/thesis/lib/python3.12/site-packages/dash_bootstrap_components/_table.py:5: UserWarning: 
The dash_html_components package is deprecated. Please replace
`import dash_html_components as html` with `from dash import html`
  import dash_html_components as html


17.3 ms ± 37 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)
({'A': 20000.0, 'B': 19999.0, 'C': 3.0}, ['C'])
142 ms ± 3.66 ms per loop (mean ± std. dev. of 7 runs, 10 loops each)
{'A': 0.0009999098850159882, 'B': 0.0009999098850160255, 'C': 2.9980001802300764}


# Findings

delegations = {
    "A": {"B": 0.9999,
          "C": 0.0001},
    "B": {"A": 1}}

- LP: 17.3 ms ± 37 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

- Iterative: 142 ms ± 3.66 ms per loop (mean ± std. dev. of 7 runs, 10 loops each), iterated 138150 times with a cutoff of 0.001

Also, when making the delegation to C smaller, LPs runtime does not change drastically.
When A->C has a weight of 10^-6, as opposed to 10^-4, the runtimes are:

- LP: 18.3 ms ± 474 μs per loop (mean ± std. dev. of 7 runs, 100 loops each)

- Iterative: 14.2 s ± 111 ms per loop (mean ± std. dev. of 7 runs, 1 loop each)




